File: desktop/ui/task_list.mjs
/**
 * Decisions behind the task list — the desktop half of what `/task` produces.
 *
 * ## Why this is its own module
 *
 * Same reason as `chat_commands.mjs` and `focus_view.mjs`: `main.js` is DOM glue
 * and cannot be unit tested, so what *decides* something lives here under
 * `node --test`. The Android twin is `android/…/ui/TaskList.kt` and the two are
 * pinned by mirrored vectors.
 *
 * ## The rule that matters
 *
 * **The buttons a row offers must mirror `karya::may_transition` exactly.** Core
 * refuses a transition the sender has no standing to make — an assigner cannot
 * mark someone else's task done, an assignee cannot withdraw it, nothing reopens
 * — and a UI that offered those buttons anyway would be showing a control that
 * returns "that is not yours to change". Offering only what will be accepted is
 * the whole job of {@link actionsFor}, and its tests are written against the same
 * table `karya.rs`'s module header states.
 *
 * Nothing here re-derives *who* you are: `TaskDto`'s `assigned_by_me` and
 * `mine_to_do` are computed once in `comrade_ui` so three frontends do not each
 * compare npubs.
 *
 * ## Field names are snake_case here and camelCase on Android, deliberately
 *
 * Tauri serialises `TaskDto` with serde and no `rename_all`, so this side reads
 * `assigned_by_me` — the same convention as `upi_intents` and `reply_to`
 * everywhere else in `main.js`. uniffi generates Kotlin properties, so
 * `TaskList.kt` reads `assignedByMe`. The two mirror each other in *behaviour*,
 * not in spelling, and getting it wrong here fails **silently**: every row would
 * simply show no buttons, because both flags would read `undefined`.
 */

/** Mark it finished. Offered to whoever was asked. */
export const DONE = "done";
/** Say no. Offered to whoever was asked, and a first-class outcome. */
export const DECLINE = "decline";
/** Take it back. Offered to whoever asked. */
export const WITHDRAW = "withdraw";

/**
 * The state strings `karya::TaskState` serialises to. Exported so a test can
 * assert this module handles every one of them rather than silently treating an
 * unknown state as open.
 */
export const STATES = ["open", "done", "declined", "withdrawn"];

/**
 * Which buttons a row may offer, mirroring `karya::may_transition`.
 *
 * | | assigner | assignee |
 * |---|---|---|
 * | withdraw | yes | no |
 * | done | no* | yes |
 * | decline | no | yes |
 *
 * *A note to self is the exception the whole table bends around: one person
 * holds both roles, so they get all three. `Task::apply` special-cases it, and
 * so does this.
 *
 * A terminal task offers nothing — nothing reopens, because "done, no wait, not
 * done" is an argument two devices would have to arbitrate.
 */
export function actionsFor(task) {
  if (!task || task.state !== "open") return [];
  const noteToSelf = task.assignee == null && task.assigned_by_me;
  if (noteToSelf) return [DONE, DECLINE, WITHDRAW];
  const out = [];
  if (task.mine_to_do) out.push(DONE, DECLINE);
  if (task.assigned_by_me) out.push(WITHDRAW);
  return out;
}

/**
 * The state a button asks core to move a task to.
 *
 * Separate from the button's own name because the two are not the same word —
 * you press "Decline" and the task becomes `declined` — and because the *casing*
 * here is a wire contract, not a style choice: `karya::TaskState` is
 * `#[serde(rename_all = "snake_case")]`, so serde rejects `"Done"` outright.
 * Sending the capitalised form made every button on this panel return an error,
 * which is why this mapping is a tested function and not a literal in an
 * onclick.
 */
export function wireState(action) {
  return { [DONE]: "done", [DECLINE]: "declined", [WITHDRAW]: "withdrawn" }[action] || null;
}

/**
 * Open tasks first (newest first), then everything resolved.
 *
 * `tasks()` already returns newest-first, so this is a partition rather than a
 * sort — re-sorting here would be a second ordering rule to keep in step with
 * the store's.
 */
export function groupTasks(tasks) {
  const list = Array.isArray(tasks) ? tasks : [];
  return {
    open: list.filter((t) => t.state === "open"),
    resolved: list.filter((t) => t.state !== "open"),
  };
}

/**
 * The line under a task saying whose it is.
 *
 * Never a raw npub if the caller can help it: `nameFor` is the caller's own
 * display-name resolver, whatever precedence that frontend already uses for
 * every other surface — `displayName` here (published handle, then a short key),
 * `peerTitle` on Android (which also consults the alias you chose). This module
 * deliberately does not pick: a task list that named people differently from the
 * chat list beside it would read as two different address books.
 *
 * It degrades to the key rather than throwing. A list showing `npub1…` is poor;
 * a list that crashes on one unknown contact is worse.
 */
export function subtitleFor(task, nameFor) {
  if (!task) return "";
  const name = (npub) => (typeof nameFor === "function" ? nameFor(npub) : npub);
  if (task.assignee == null && task.assigned_by_me) return "yours";
  if (task.assigned_by_me) return `you asked ${name(task.assignee)}`;
  return `${name(task.assigner)} asked you`;
}

/** How a resolved task reads. Open tasks say nothing — the row is the point. */
export function stateLabel(state) {
  return { done: "Done", declined: "Declined", withdrawn: "Withdrawn" }[state] || "";
}

/**
 * What an empty list should say.
 *
 * Names the command, because a screen reachable before you know the command
 * exists is the one place to teach it — and because "no tasks" teaches nothing.
 */
export function emptyCopy() {
  return 'Nothing yet. Type "/task water the plants" in any chat — or "/task ship it @ana" to ask a comrade.';
}

File: android/app/src/main/java/mullu/comrade/ui/TaskList.kt
package mullu.comrade.ui

import mullu.comrade.ComradeCore
import uniffi.comrade_core.TaskState

/**
 * Decisions behind the task list — the Android half of what `/task` produces.
 *
 * Compose-free and Android-free so the whole thing is unit-testable on a plain
 * JVM, like [ComposerMode], [ChatThread] and [ChatCommands]. The desktop twin is
 * `desktop/ui/task_list.mjs`, pinned by mirrored vectors.
 *
 * ## The rule that matters
 *
 * **The buttons a row offers mirror `karya::may_transition` exactly.** Core
 * refuses a transition the caller has no standing to make — an assigner cannot
 * mark someone else's task done, an assignee cannot withdraw it, nothing reopens
 * — so a UI that offered those anyway would show a control that comes back with
 * "that is not yours to change". Offering only what will be accepted is the
 * whole job of [actionsFor].
 *
 * Nothing here re-derives *who* you are: `TaskInfo.assignedByMe` and `mineToDo`
 * are computed once in `comrade_ui` so three frontends do not each compare npubs.
 */

/** What a row can offer. */
enum class TaskAction {
    /** Mark it finished. For whoever was asked. */
    Done,

    /** Say no — a first-class outcome, not a failure. For whoever was asked. */
    Decline,

    /** Take it back. For whoever asked. */
    Withdraw,
}

/** A task list split for rendering. */
data class GroupedTasks(
    val open: List<ComradeCore.TaskInfo>,
    val resolved: List<ComradeCore.TaskInfo>,
)

object TaskList {

    /**
     * Which buttons a row may offer, mirroring `karya::may_transition`:
     *
     * | | assigner | assignee |
     * |---|---|---|
     * | withdraw | yes | no |
     * | done | no* | yes |
     * | decline | no | yes |
     *
     * *A note to self is the exception the table bends around: one person holds
     * both roles, so they get all three. `Task::apply` special-cases it and so
     * does this — refusing to let somebody delete their own private note would
     * be absurd.
     *
     * A task that has already landed offers nothing; nothing reopens.
     */
    fun actionsFor(task: ComradeCore.TaskInfo): List<TaskAction> {
        if (task.state != TaskState.OPEN) return emptyList()
        val noteToSelf = task.assignee == null && task.assignedByMe
        if (noteToSelf) {
            return listOf(TaskAction.Done, TaskAction.Decline, TaskAction.Withdraw)
        }
        return buildList {
            if (task.mineToDo) {
                add(TaskAction.Done)
                add(TaskAction.Decline)
            }
            if (task.assignedByMe) add(TaskAction.Withdraw)
        }
    }

    /**
     * Open first, resolved after, each keeping the order `tasks()` returned.
     *
     * A partition rather than a sort: the store already answers newest-first, and
     * re-sorting here would be a second ordering rule to keep in step with it.
     */
    fun group(tasks: List<ComradeCore.TaskInfo>): GroupedTasks = GroupedTasks(
        open = tasks.filter { it.state == TaskState.OPEN },
        resolved = tasks.filter { it.state != TaskState.OPEN },
    )

    /**
     * The line under a task saying whose it is.
     *
     * Never a raw npub if the caller can help it: [nameFor] is the caller's own
     * display-name resolver, so this list names people exactly the way every
     * other screen does — `TaskListScreen` passes `peerTitle`, which is the
     * alias you chose, then the handle they published, then a short key. This
     * function deliberately does not pick that order itself; a task list naming
     * people differently from the chat list would read as a second address book.
     */
    fun subtitleFor(task: ComradeCore.TaskInfo, nameFor: (String) -> String): String = when {
        task.assignee == null && task.assignedByMe -> "yours"
        task.assignedByMe -> "you asked ${nameFor(task.assignee ?: "")}"
        else -> "${nameFor(task.assigner)} asked you"
    }

    /** How a resolved task reads. An open one says nothing — the row is the point. */
    fun stateLabel(state: TaskState): String = when (state) {
        TaskState.OPEN -> ""
        TaskState.DONE -> "Done"
        TaskState.DECLINED -> "Declined"
        TaskState.WITHDRAWN -> "Withdrawn"
    }

    /**
     * What an empty list says.
     *
     * Names the command, because this screen is reachable before you know `/task`
     * exists and is therefore the one place that can teach it. "No tasks" teaches
     * nothing.
     */
    const val EMPTY_COPY: String =
        "Nothing yet. Type \"/task water the plants\" in any chat — " +
            "or \"/task ship it @ana\" to ask a comrade."
}
